src/material.rs

```diff
@@ -0,0 +1,353 @@
+}
+
+/// A uniform pigment.
+///
+/// This is the most boring pigment: a uniform hue over the whole surface.
+#[derive(Default)]
+pub struct UniformPigment {
+    /// A [`Color`].
+    pub color: Color,
+}
+
+impl GetColor for UniformPigment {
+    fn get_color(&self, _uv: Vector2D) -> Color {
+        self.color
+    }
+}
+
+/// A textured pigment.
+///
+/// The texture is given through a [`HdrImage`], maybe read from pfm file.
+pub struct ImagePigment<'a> {
+    /// An [`HdrImage`] reference.
+    hdr_img: &'a HdrImage,
+}
+
+impl<'a> ImagePigment<'a> {
+    /// Create a new [`ImagePigment`] from [`HdrImage`].
+    pub fn new(hdr_img: &'a HdrImage) -> Self {
+        Self { hdr_img }
+    }
+}
+
+impl<'a> GetColor for ImagePigment<'a> {
+    fn get_color(&self, uv: Vector2D) -> Color {
+        let mut col = (uv.u * self.hdr_img.shape().0 as f32) as u32;
+        let mut row = (uv.v * self.hdr_img.shape().1 as f32) as u32;
+        if col >= self.hdr_img.shape().0 {
+            col = self.hdr_img.shape().0 - 1;
+        }
+        if row >= self.hdr_img.shape().1 {
+            row = self.hdr_img.shape().1 - 1
+        }
+        // TODO
+        // A nicer solution would implement bilinear interpolation
+        // to reduce pixelization artifacts.
+        // See https://en.wikipedia.org/wiki/Bilinear_interpolation.
+        self.hdr_img.get_pixel(col, row).unwrap()
+    }
+}
+
+/// A checkered pigment.
+///
+/// The number of rows/columns in the checkered pattern is tunable,\
+/// but you cannot have a different number of repetitions along the u/v directions.
+pub struct CheckeredPigment {
+    /// First [`Color`].
+    pub color1: Color,
+    /// Second [`Color`].
+    pub color2: Color,
+    /// Number of steps.
+    pub steps: u32,
+}
+
+impl GetColor for CheckeredPigment {
+    fn get_color(&self, uv: Vector2D) -> Color {
+        let int_u = f32::floor(uv.u * self.steps as f32) as u32;
+        let int_v = f32::floor(uv.v * self.steps as f32) as u32;
+        if int_u % 2 == int_v % 2 {
+            self.color1
+        } else {
+            self.color2
+        }
+    }
+}
+
+/// A trait for evaluating a particular BRDF on a parametric surface `(u,v)`.
+pub trait Eval {
+    fn eval(&self, normal: Normal, in_dir: Vector, out_dir: Vector, uv: Vector2D) -> Color;
+}
+
+/// A trait for scatter a [`Ray`] for a particular BRDF.
+pub trait ScatterRay {
+    fn scatter_ray(
+        &self,
+        pcg: &mut Pcg,
+        incoming_dir: Vector,
+        interaction_point: Point,
+        normal: Normal,
+        depth: u32,
+    ) -> Ray;
+}
+
+/// A class representing an ideal diffuse BRDF (also called "Lambertian").
+pub struct DiffuseBRDF<P: GetColor> {
+    /// A generic pigment that implement [`GetColor`].
+    pub pigment: P,
+}
+
+impl Default for DiffuseBRDF<UniformPigment> {
+    fn default() -> Self {
+        Self {
+            pigment: UniformPigment { color: WHITE },
+        }
+    }
+}
+
```

Comment:
- is white a good choice as default color? maybe a nicer color would help during debugging.
- why `DiffuseBRDF<CheckeredPigment> doesn't have a `Default` implementation?